### PR TITLE
Add initialization callbacks to plugins

### LINKED ISSUE #2750: [plugin] Add trace processor before exporter
By adding a customized trace processor before the exporter in SDK, we can have a chance to check each Activity and decide if it is going to be exported. However, currently plugin is handled after [UseEnvironmentVariables()](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/36dd368fc186c165e7e2a36ec0455aef99ad01a7/src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs#L119), so the processor added as below is always after the exporter. 
```
public class Plugin
{
    public TracerProviderBuilder ConfigureTracerProvider(TracerProviderBuilder builder)
    {
        return builder.AddProcessor(new MyProcessor());
    }
}
```
can we do sth in plugin at the very beginning of CreateTracerProviderBuilder ?
Discussions can be found in slack: https://cloud-native.slack.com/archives/C01NR1YLSE7/p1689279656814439

Another request is can we do sth in plugin after [TracerProvider ](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/36dd368fc186c165e7e2a36ec0455aef99ad01a7/src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs#L122C1-L122C1) got build or at the end of [Initialize()](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/36dd368fc186c165e7e2a36ec0455aef99ad01a7/src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs#L66C4-L66C4). Sth similar to [_pluginManager.Initializing()](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/36dd368fc186c165e7e2a36ec0455aef99ad01a7/src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs#L96), but it is `_pluginManager.Initialized()`. We may start to collect some metrics that are related to the TracerProvider etc

### PR DESCRIPTION
**Pull request by RassK**
## Why

Fixes #2750 (Part 2)

## What

Adds initialization events:
* `TraceProviderInitialized(TracerProvider provider)`
* `MeterProviderInitialized(MeterProvider provider)`

## Tests

Extended unit tests to cover additional methods.

## Checklist

- [x] `CHANGELOG.md` is updated.
- [x] Documentation is updated.
- [x] New features are covered by tests.

### CODE REVIEW

**Review by nrcventura — APPROVED** [review]

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Plugins/PluginManager.cs [hunk 1]
```diff
@@ -100,61 +105,60 @@ public MeterProviderBuilder AfterConfigureMeterProviderBuilder(MeterProviderBuil
...
+        return builder;
     }
 
-    private void ConfigureOptions(Type type, object options, string methodName)
+    private void CallPlugins(string methodName)
     {
         foreach (var plugin in _plugins)
         {
-            var mi = plugin.Type.GetMethod(methodName, new[] { type });
+            var mi = plugin.Type.GetMethod(methodName, Type.EmptyTypes);
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Plugins/PluginManager.cs:145]: nice

**Review by pjanotti — COMMENTED** [review]
LGTM, with 2 minor nits

File: docs/plugins.md [hunk 2]
```diff
@@ -18,6 +18,18 @@ public class MyPlugin
         // My custom logic here
     }
 
+    // To access TraceProvider right after TraceProviderBuilder.Build() is executed.
+    public void TraceProviderInitialized(TracerProvider tracerProvider)
```
[the remark this answers is not in the document]
Comment (RassK) [reply docs/plugins.md]: That's excellent catch, there were so many of those https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2760/commits/b8ae99bd2a21b6c5806f120cf5c4aa63fbb7a5ef

File: src/OpenTelemetry.AutoInstrumentation/Plugins/PluginManager.cs [hunk 3]
```diff
@@ -41,16 +41,11 @@ public PluginManager(GeneralSettings settings)
         _plugins = plugins;
     }
 
+    internal IReadOnlyList<(Type Type, object Instance)> Plugins => _plugins;
```
[the remark this answers is not in the document]
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Plugins/PluginManager.cs:44]: There are hard references in Visual Studio directly to the test method. I would have added if it was reflection. Do you think this is enough?

[the remark this answers is not in the document]
Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation/Plugins/PluginManager.cs:44]: You are right that reading it from an IDE like VS or Rider you can easily see how the property is being used. However, a comment shows that it was actually created with that intent (to be used in tests) in mind.

[the remark this answers is not in the document]
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Plugins/PluginManager.cs:45]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2760/commits/d7cb543c14405e8f89c154d307c52ebf26d082dd

**Review by pjanotti — APPROVED** [review]

**Pull request merged**